lib/rpmsg/rpmsg_virtio.c

```diff
@@ -355,7 +356,7 @@ static int rpmsg_virtio_send_offchannel_nocopy(struct rpmsg_device *rdev,
 	hdr = RPMSG_LOCATE_HDR(data);
 	/* The reserved field contains buffer index */
 	idx = hdr->reserved;
-
+#if 0
```

Comment:
The test can't pass the test after I change the rpmsg_virtio_send_offchannel_raw to reuse the nocopy API, I am debugging this problem, so please ignore this PR until I find the root cause.

Reply:
I tried these patches and found a problem which I think also is the root cause for your segment fault.

Following line, inside the function "rpmsg_virtio_send_offchannel_nocopy" causes the problem:
buff_len = virtqueue_get_buffer_length(rvdev->svq, idx);

virtqueue_get_buffer_length reads buffer length from the buffer descriptor. When a buffer is allocated by rpmsg_virtio_shm_pool_get_buffer (inside function rpmsg_virtio_get_tx_buffer) instead of from used-ring, reading the length of the buffer from the buffer descriptor is not correct. As a result, whenever a Master sends the first message, the length of the buffer return by this function is observed to be ZERO. Which leads to a segment fault in subsequent message sends.

As a temporary fix, you can replace the above line with the following line. There should not be a segment fault after this change.
buff_len = RPMSG_BUFFER_SIZE;

I am also putting a comment on the review to trek this mistake.

Reply:
Thanks, @kr-satish, I will try your suggestion soon.